activerecord/lib/active_record/connection_adapters/abstract/query_cache.rb

```diff
@@ -123,6 +123,7 @@ def cache_sql(sql, name, binds)
         # If arel is locked this is a SELECT ... FOR UPDATE or somesuch. Such
         # queries should not be cached.
         def locked?(arel)
+          arel = arel.arel if arel.is_a?(Relation)
           arel.respond_to?(:locked) && arel.locked
```

Comment:
At least statement cache will pass compiled SQL string to `find_by_sql`.
https://github.com/rails/rails/blob/v5.1.2/activerecord/lib/active_record/statement_cache.rb#L107